### PR TITLE
Guidelines for the use of eBPF in OCP

### PR DESCRIPTION
**Pull request by danwinship**
People want to use eBPF in OCP. This is awesome and scary, and we need to come up with guidelines on how this will work.

This is currently very very WIP and I'll need help from various other teams to fill in all of the details.

### CODE REVIEW

File: enhancements/ebpf-in-ocp.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+### User Stories
+
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
+using eBPF in OCP.)
+
+#### Upcoming Network Observability Features
+
+(Something about upcoming network observability features.)
```
Comment (mariomac) [inline enhancements/ebpf-in-ocp.md:45]:
```suggestion
As an OCP developer, I want to attach the eBPF agent into different parts of
the system (e.g. Nodes' kernel) to be able to trace the network flows within
the system, as well as being able to correlate them with other parts of the
system configuration (e.g. Network Policies, security, etc.).
```

Comment (jotak) [reply enhancements/ebpf-in-ocp.md:45]: should we also describe the end user (OCP customer) use cases that we address? Something like: "as an OCP customer, I want to be able to trace network flows across pods and services in my cluster, with a resource overhead as low as possible." ?

Comment (danwinship) [reply enhancements/ebpf-in-ocp.md:45]: I don't think we need to get into the higher-level goals that led to people using eBPF... for purposes of this enhancement, we're starting with "I want to use eBPF to do X" and not really worrying about how you got to that point. At least for OCP-internal use cases. For customer use cases we may need to distinguish between "customer wants to use eBPF and we think it's reasonable" vs "customer wants to use eBPF and we think it's a bad idea". (eg, that's why I separated out the "eBPF for bugfixes" and "eBPF for new features" use cases, because we might have different opinions on them)

File: enhancements/ebpf-in-ocp.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+replaces:
+superseded-by:
+---
+
+# Guidelines for the Use of eBPF in OCP
+
+## Summary
```
Comment (eranra) [inline enhancements/ebpf-in-ocp.md:25]: @danwinship should add something on listing and addressing support OCP platforms. So that eBPF programs are tested and documented as supported for a known set of OCP architectures (x86,arm ... )

Comment (eranra) [reply enhancements/ebpf-in-ocp.md:25]: @danwinship same with regards to versioning and required kernel levels... so that it will be possible to list dynamically all supported loaded eBPF programs and also address the issue of upgrading etc.

File: enhancements/ebpf-in-ocp.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+on a kernel bond interface, using eBPF. The upstream Linux maintainer
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
```
Comment (astoycos) [inline enhancements/ebpf-in-ocp.md:83]: I feel like CI will be the best way to ensure this... should we have an "ebpf feature" CI workflow to ensure stable cooperation? Maybe a job running with cilium + a job running with any OCP specific ebpf features

Comment (astoycos) [reply enhancements/ebpf-in-ocp.md:83]: (obviously good documentation on what hooks we're using explicitly is important to) but yeah the problem of ebpf interoperability is definitely a complex one.

File: enhancements/ebpf-in-ocp.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+As an OCP developer/product manager, I would like to introduce
+functionality in a future release that makes use of the XDP and/or tc
+eBPF hooks to modify data plane functionality, and have these new
+features interact in a predictable way with existing networking
+functionality, other operators, and local customizations.
+
+#### Security
+
+As a cluster administrator, I would like to make sure that no one is
```
Comment (astoycos) [inline enhancements/ebpf-in-ocp.md:102]: Currently only privileged pods would be able to run ebpf programs... So the admin should know about any of these right?  Also As we start running bpf programs it probably will make sense to have an OCP system pod that gives some stats on all of the bpf programs currently running in a cluster (maybe this is observability's Job)

File: enhancements/ebpf-in-ocp.md (hunk 3 above)
Comment (vrutkovs) [reply enhancements/ebpf-in-ocp.md:83]: We do have Cilium jobs in CI - https://prow.ci.openshift.org/?job=*cilium*. They however run just the conformance suite, no custom eBPF tests

File: enhancements/ebpf-in-ocp.md (hunk 4 above)
Comment (mariomac) [reply enhancements/ebpf-in-ocp.md:102]: Being more precise, it is not mandatory that pods are full-privileged to run eBPF programs. For example, the NetObserv eBPF agent runs in a non-privileged pod, in a container defining the following security context:

```yaml
        securityContext:
          capabilities:
            add:
              - BPF
              - PERFMON
              - NET_ADMIN
              - SYS_RESOURCE
          runAsUser: 0
```

**Comment by williamcaban** [comment]
A use case for consideration: 
- There is a lot of interest from partners and customers for reducing their dependencies to DPDK while still getting "good enough" performance using eBPF based stack. Also, be able to do that without secondary interfaces. E.g. have the capability native on the primary Pod interface (eth0).
- Alternatives to a vhost-user and virtio-pmd, etc

File: enhancements/ebpf-in-ocp.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
+using eBPF in OCP.)
+
+#### Upcoming Network Observability Features
+
+(Something about upcoming network observability features.)
+
+#### Ensuring Network (and General) Supportability
```
Comment (dofinn) [inline enhancements/ebpf-in-ocp.md:47]: This would be amazing for CEE supporting managed OpenShift

File: enhancements/ebpf-in-ocp.md (hunk 3 above)
Comment (dofinn) [reply enhancements/ebpf-in-ocp.md:83]: additionally there is pixie https://docs.px.dev/about-pixie/pixie-ebpf/

File: enhancements/ebpf-in-ocp.md (hunk 4 above)
Comment (sherine-k) [reply enhancements/ebpf-in-ocp.md:102]: Would SELinux profile (CAP_BPF) help keep the attack surface a little smaller than using privileged containers (when needed)?

File: enhancements/ebpf-in-ocp.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+eBPF hooks to modify data plane functionality, and have these new
+features interact in a predictable way with existing networking
+functionality, other operators, and local customizations.
+
+#### Security
+
+As a cluster administrator, I would like to make sure that no one is
+using eBPF that I don't know about, because I am in a highly regulated
+industry, and I can't have random people running mystery code in my
+kernel, even if it's sandboxed.
```
Comment (netoptimizer) [inline enhancements/ebpf-in-ocp.md:105]: One option for this "cluster administrator" would be to write a BPF LSM hook program that denies loading BPF programs regardless of their security level/capabilities.
@tohojo wrote this example to disable BPF:
https://github.com/xdp-project/bpf-examples/blob/master/lsm-nobpf

File: enhancements/ebpf-in-ocp.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+use eBPF and how we expect others to use (or not use it) to prevent
+conflicts later.
+
+## Motivation
+
+### User Stories
+
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
```
Comment (dave-tucker) [inline enhancements/ebpf-in-ocp.md:40]: As an OCP customer I want to orchestrate one (or more) probes to run on my nodes for either monitoring application performing application debugging.

As an OCP customer who is unable to modify application code, I would like to add eBPF-based instrumentation for my application.

**Comment by saschagrunert** [comment]
I think this would be still valuable
/reopen

File: enhancements/ebpf-in-ocp.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+### Drawbacks
+
+None. We really need guidelines here.
+
+## Design Details
+
+### Open Questions [optional]
+
+Many
```
Comment (dave-tucker) [inline enhancements/ebpf-in-ocp.md:159]: Open question:

Where do we stand with support on RHEL. The [latest release notes](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/9.0_release_notes/new-features) state:

>  The XDP and AF_XDP API parts of the libbpf library are not supported and may be removed in a future release. 

We may not use `libbpf` to load XDP or AF_XDP programs.

> The eXpress Data Path (XDP) feature, which provides access to the received packets before the kernel networking stack processes them. Red Hat supports XDP only if it is used through the libxdp library.

We must use libxdp BUT we do not have higher level language bindings for it (i.e for Go).

> The xdp-tools package, which contains user-space support utilities for the XDP feature and is supported on the AMD64 and Intel64 CPU architectures.

libxdp is not packaged for any architecture other than x86_64

Comment (tohojo) [reply enhancements/ebpf-in-ocp.md:159]: For RHEL9, libxdp is packaged for all archs, it just won't do multi-prog until the kernel support lands; but that's fine, it'll fall back automatically, so it should be totally fine to use it as-is for those archs. So the lack of Go bindings is the only blocker AFAICT...

Comment (dave-tucker) [reply enhancements/ebpf-in-ocp.md:159]: thanks for clarifying @tohojo
we'll have to figure out a design for libxdp-go that works seamlessly with `cillium/ebpf`
lets' chat about that next week and perhaps we can get something started

File: enhancements/ebpf-in-ocp.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
+Cilium uses a wide variety of eBPF hooks, including FIXME.
```
Comment (markmc) [inline enhancements/ebpf-in-ocp.md:84]: This talk on Cilium's experience seems relevant: https://lpc.events/event/16/contributions/1353/attachments/1068/2130/plumbers_2022_tc_bpf_links.pdf

tl;dr - they saw datadog stomping on Cilium's hooks in a customer environment; they believe bpf_link attachment semantics help with this problem; they propose a model for using bpf_link with tc

Comment (netoptimizer) [reply enhancements/ebpf-in-ocp.md:84]: Yes, @markmc this example of (Datadog) BPF stomping on (Cilium) BPF is exactly the problem I'm trying to highlight.

For legal/support reasons RH must document what BPF hooks our products use and state RH cannot support shared usage of said BPF hooks. Regardless this stomping _will occur in practice_ in production environments, leading to **hard-to-debug** scenarios.

**IMHO** we should implement a technical solution, that:

1. as minimum detects and reports if a products BPF hook get "stomped"
2. preferably detects and _denies_ others from taking over products BPF hook
